test/runner/run.js

```diff
@@ -0,0 +1,337 @@
+import chalk from "chalk";
+import { asyncExitHook, gracefulExit } from "exit-hook";
+import { getLatestBrowser } from "./browserstack/api.js";
+import { buildBrowserFromString } from "./browserstack/buildBrowserFromString.js";
+import { localTunnel } from "./browserstack/local.js";
+import { reportEnd, reportTest } from "./reporter.js";
+import { createTestServer } from "./createTestServer.js";
+import { buildTestUrl } from "./lib/buildTestUrl.js";
+import { generateHash, printModuleHashes } from "./lib/generateHash.js";
+import { getBrowserString } from "./lib/getBrowserString.js";
+import { cleanupAllJSDOM, cleanupJSDOM } from "./jsdom.js";
+import { modules as allModules } from "./modules.js";
+import { cleanupAllBrowsers, touchBrowser } from "./browserstack/browsers.js";
+import {
+	addBrowserStackRun,
+	getNextBrowserTest,
+	retryTest,
+	runAllBrowserStack
+} from "./browserstack/queue.js";
+import { addSeleniumRun, runAllSelenium } from "./selenium/queue.js";
+
+const EXIT_HOOK_WAIT_TIMEOUT = 60 * 1000;
+
+/**
+ * Run modules in parallel in different browser instances.
+ */
+export async function run( {
+	amd,
+	browsers: browserNames,
+	browserstack,
+	concurrency,
+	debug,
+	headless,
+	isolate,
+	modules = [],
+	retries = 0,
+	runId,
+	verbose
+} ) {
+	if ( !browserNames || !browserNames.length ) {
+		browserNames = [ "chrome" ];
+	}
+	if ( !modules.length ) {
+		modules = allModules;
+	}
+	if ( headless && debug ) {
+		throw new Error(
+			"Cannot run in headless mode and debug mode at the same time."
+		);
+	}
+
+	if ( verbose ) {
+		console.log( browserstack ? "Running in BrowserStack." : "Running locally." );
+	}
+
+	const errorMessages = [];
+	const pendingErrors = {};
+
+	// Convert browser names to browser objects
+	let browsers = browserNames.map( ( b ) => ( { browser: b } ) );
+	const tunnelId = generateHash(
+		`${ Date.now() }-${ modules.join( ":" ) }-${ ( browserstack || [] )
+			.concat( browserNames )
+			.join( ":" ) }`
+	);
+
+	// A unique identifier for this run
+	if ( !runId ) {
+		runId = tunnelId;
+	}
+
+	// Create the test app and
+	// hook it up to the reporter
+	const reports = Object.create( null );
+	const app = await createTestServer( ( message ) => {
+		switch ( message.type ) {
+			case "testEnd": {
+				const reportId = message.id;
+				const report = reports[ reportId ];
+				touchBrowser( report.browser );
+				const errors = reportTest( message.data, reportId, report );
+				pendingErrors[ reportId ] ||= {};
```

Comment:
How about two changes here:
```suggestion
				pendingErrors[ reportId ] ??= Object.create( null );
```
Since keys are possibly arbitrary from the perspective of this code: `message.data.name` - `Object.create( null )` seems safer.